sdk/eventhub/azure-eventhub/azure/eventhub/_consumer.py

```diff
@@ -216,21 +209,20 @@ def receive(self, batch=False, max_batch_size=300, max_wait_time=None):
         self._receive_start_time = self._receive_start_time or time.time()
         deadline = self._receive_start_time + (max_wait_time or 0)  # max_wait_time can be None
         if len(self._message_buffer) < max_batch_size:
+            # TODO: the retry here is a bit tricky as we are using low-level api from the amqp client.
+            #  besides, the retry will create a new AMQPClient + new link properties (offset)
+            #  so we need to create new amqp client or update the link properties of existing one
+            #  I'd like to add more optional steps into the retry policy of amqp lib
+            #  to allow updating existing properties for recreation of a client if link/connection error happens
             while retried_times <= max_retries:
                 try:
                     if self._open():
-                        messages =self._handler.receive_message_batch(
-                            max_batch_size=max_batch_size
-                        )
-                        for message in messages:
-                            uamqp_message = Message(**message)
-                            self._message_buffer.appendleft(uamqp_message)
-                        #self._handler.do_work()  # type: ignore
+                        self._handler.do_work(batch=self._prefetch)  # type: ignore
```

Comment:
It looks like we're changing the EH Consumer to work based on callbacks rather than the receive_batch function. What was the motivation for this?

Reply:
it's not changed, current EH consumer is also based on callbacks